File: codegen/codegen-erlang/src/test/resources/ir/http_checksum_helpers.expected.erl
headers_set(Name, Value, Headers) -> lists:keystore(Name, 1, Headers, {Name, Value}).

checksum_header_encode(Data) when is_binary(Data) -> base64:encode(Data).

md5_hash(Body) -> crypto:hash(md5, Body).

sha256_hash(Body) -> crypto:hash(sha256, Body).

crc32_hash(Body) -> <<(erlang:crc32(Body)):32/big-unsigned-integer>>.

crc32c_hash(Body) -> crypto:hash(crc32c, Body).

crc64nvme_hash(_Body) -> error({unsupported_checksum_algorithm, crc64nvme}).

xxhash64_hash(_Body) -> error({unsupported_checksum_algorithm, xxhash64}).

xxhash3_hash(_Body) -> error({unsupported_checksum_algorithm, xxhash3}).

xxhash128_hash(_Body) -> error({unsupported_checksum_algorithm, xxhash128}).

checksum_digest(Body, <<"MD5">>) -> md5_hash(Body);
checksum_digest(Body, <<"SHA256">>) -> sha256_hash(Body);
checksum_digest(Body, <<"CRC32">>) -> crc32_hash(Body);
checksum_digest(Body, <<"CRC32C">>) -> crc32c_hash(Body).

validate_response_checksum(_Body, _Headers, []) ->
    ok;
validate_response_checksum(Body, Headers, [HeaderName | Rest]) ->
    case proplists:get_value(HeaderName, Headers, undefined) of
        undefined ->
            validate_response_checksum(Body, Headers, Rest);
        Expected ->
            case
                checksum_header_encode(
                    checksum_digest(Body, checksum_algorithm_from_header(HeaderName))
                ) =:= Expected
            of
                true -> ok;
                false -> {error, {checksum_mismatch, HeaderName}}
            end
    end.

checksum_algorithm_from_header(<<"x-amz-checksum-", Rest/binary>>) ->
    list_to_binary(string:uppercase(binary_to_list(Rest))).

File: codegen/codegen-elixir/src/test/resources/ir/http_checksum_helpers.expected.ex
defp headers_set(name, value, headers), do: List.keystore(name, 0, headers, {name, value})

defp checksum_header_encode(data) when is_binary(data), do: Base.encode64(data)

defp md5_hash(body), do: :crypto.hash(:md5, body)

defp sha256_hash(body), do: :crypto.hash(:sha256, body)

defp crc32_hash(body), do: <<:erlang.crc32(body) :: "32-big-unsigned-integer">>

defp crc32c_hash(body), do: :crypto.hash(:crc32c, body)

defp checksum_digest(body, "MD5"), do: md5_hash(body)
defp checksum_digest(body, "SHA256"), do: sha256_hash(body)
defp checksum_digest(body, "CRC32"), do: crc32_hash(body)
defp checksum_digest(body, "CRC32C"), do: crc32c_hash(body)

defp validate_response_checksum(_body, _headers, []), do: :ok
defp validate_response_checksum(
  body,
  headers,
  [header_name | rest]
) do
  List.keyfind(headers, header_name, 0)
  |> case do
    {_, expected} -> validate_checksum_match(body, header_name, expected)
    nil -> validate_response_checksum(body, headers, rest)
  end
end

defp validate_checksum_match(
  body,
  header_name,
  expected
) do
  algorithm = checksum_algorithm_from_header(header_name)
  computed = checksum_header_encode(checksum_digest(body, algorithm))
  if(computed == expected, do: :ok, else: {:error, {:checksum_mismatch, header_name}})
end

defp checksum_algorithm_from_header(header_name), do: String.upcase(String.replace_prefix(header_name, "x-amz-checksum-", ""))